File: easy/3912_valid_elements_in_an_array/solution.cpp
#include <vector>

class Solution {
public:
  std::vector<int> findValidElements(std::vector<int> &nums) {
    const size_t numsLen = nums.size();

    std::vector<int> prefixMaxes(numsLen, 0);
    std::vector<int> suffixMaxes(numsLen, 0);

    for (size_t i = 1, j = numsLen - 2; i < numsLen; ++i, --j) {
      prefixMaxes[i] = std::max(prefixMaxes[i - 1], nums[i - 1]);
      suffixMaxes[j] = std::max(suffixMaxes[j + 1], nums[j + 1]);
    }

    std::vector<int> validNums = {nums[0]};

    for (size_t i = 1; i < numsLen - 1; ++i) {
      if (nums[i] > prefixMaxes[i] || nums[i] > suffixMaxes[i])
        validNums.push_back(nums[i]);
    }

    if (numsLen != 1)
      validNums.push_back(nums.back());

    return validNums;
  }
};

File: easy/3912_valid_elements_in_an_array/solution.py
from typing import List


class Solution:
    def findValidElements(self, nums: List[int]) -> List[int]:
        prefix_maxes = [0] * len(nums)
        suffix_maxes = [0] * len(nums)

        for i in range(1, len(nums)):
            prefix_maxes[i] = max(prefix_maxes[i - 1], nums[i - 1])
        for i in range(len(nums) - 2, -1, -1):
            suffix_maxes[i] = max(suffix_maxes[i + 1], nums[i + 1])

        valid_nums = [nums[0]]

        for i in range(1, len(nums) - 1):
            if nums[i] > prefix_maxes[i] or nums[i] > suffix_maxes[i]:
                valid_nums.append(nums[i])

        if len(nums) != 1:
            valid_nums.append(nums[-1])

        return valid_nums
